fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/impl/registry/sync/PersistentDynamicRegistryHandler.java

```diff
@@ -0,0 +1,200 @@
+	}
+
+	@NotNull
+	private static CompoundTag remapDynamicRegistries(DynamicRegistryManager.Impl dynamicRegistryManager, @Nullable CompoundTag existingTag) throws RemapException {
+		CompoundTag registries = new CompoundTag();
+
+		// For now we only care about biomes, but lets keep our options open
+		CompoundTag biomeRegistryData = null;
+
+		if (existingTag != null) {
+			biomeRegistryData = existingTag.getCompound(Registry.BIOME_KEY.getValue().toString());
+		}
+
+		MutableRegistry<?> registry = dynamicRegistryManager.get(Registry.BIOME_KEY);
+		CompoundTag biomeIdMap = remapRegistry(Registry.BIOME_KEY.getValue(), registry, biomeRegistryData);
+		registries.put(Registry.BIOME_KEY.getValue().toString(), biomeIdMap);
+
+		CompoundTag outputTag = new CompoundTag();
+		outputTag.putInt("version", 1);
+		outputTag.put("registries", registries);
+		return registries;
+	}
+
+	/**
+	 * Remaps a registry if existing data is passed in.
+	 * Then writes out the ids in the registry (remapped or a new world).
+	 * Keeps hold of the orphaned registry entries as to not overwrite them.
+	 */
+	private static <T> CompoundTag remapRegistry(Identifier registryId, MutableRegistry<T> registry, @Nullable CompoundTag existingTag) throws RemapException {
+		if (!(registry instanceof RemappableRegistry)) {
+			throw new UnsupportedOperationException("Cannot remap un re-mappable registry: " + registryId.toString());
+		}
+
+		boolean isModded = registry.getIds().stream().anyMatch(id -> !id.getNamespace().equals("minecraft"));
+
+		// The current registry might not be modded, but we might have previous changed vanilla ids that we should try and remap
+		if (existingTag != null && !isModded) {
+			isModded = existingTag.getKeys().stream()
+					.map(existingTag::getString)
+					.map(Identifier::new)
+					.anyMatch(id -> !id.getNamespace().equals("minecraft"));
+		}
+
+		if (LOGGER.isDebugEnabled()) {
+			if (existingTag == null) {
+				LOGGER.debug("No existing data found, assuming new registry with {} entries. modded = {}", registry.getIds().size(), isModded);
+			} else {
+				LOGGER.debug("Existing registry data found. modded = {}", isModded);
+
+				for (T entry : registry) {
+					//noinspection unchecked
+					Identifier id = registry.getId(entry);
+					int rawId = registry.getRawId(entry);
+
+					if (id == null) continue;
+
+					if (existingTag.getKeys().contains(id.toString())) {
+						int existingRawId = existingTag.getInt(id.toString());
+
+						if (rawId != existingRawId) {
+							LOGGER.debug("Remapping {} {} -> {}", id.toString(), rawId, existingRawId);
+						} else {
+							LOGGER.debug("Using existing id for {} {}", id.toString(), rawId);
+						}
+					} else {
+						LOGGER.debug("Found new registry entry {}", id.toString());
+					}
+				}
+			}
+		}
+
+		// If we have some existing ids and the registry contains modded/datapack entries we remap the registry with those
+		if (existingTag != null && isModded) {
```

Comment:
It may be desirable to fix the bug with vanilla datapacks too?

Reply:
`isModded` in this case means any registry entry that contains an entry not in the minecraft namespace. The datapack I was testing with used its own namespace. I might be wrong but I would assume all custom biomes to not use the minecraft namespace?

Reply:
Right, that's fine IMO. Maybe clarify the comment a little?